### PR TITLE
Revert batched pdist, improve existing kernel, add test

### PR DESCRIPTION
**Pull request by gchanan**
1) Reverts https://github.com/pytorch/pytorch/pull/12302 which added support for batched pdist. Except I kept the (non-batched) test improvements that came with that PR, because they are nice to have.  Motivation: https://github.com/pytorch/pytorch/issues/15511
2) For the non-batched pdist, improved the existing kernel by forcing fp64 math and properly checking cuda launch errors
3) Added a 'large tensor' test that at least on my machine, fails on the batch pdist implementation.

### CODE REVIEW

**Comment by gchanan** [comment]
CC @erikbrinkman @ngimel.  It would be better to improve the batched kernel, of course, but that looked more involved, so I wanted to unblock this.

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 1]
```diff
@@ -79,19 +80,18 @@ struct dists {
...
-  const int l = blockIdx.x;
-  const int k = blockIdx.y;
-  const int stride = blockDim.y;
+  const int k = blockIdx.x;
+  const int stride = blockDim.x;
 
-  float n2 = n - .5;
+  double n2 = n - .5;
   // The -1 accounts for floating point truncation issues
   int64_t i = static_cast<int64_t>((n2 - device_sqrt<scalar_t>(n2 * n2 - 2 * k - 1)));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: I'd feel slightly safer with device_sqrt<double> here, but may be it's all right.

File: test/test_torch.py [hunk 2]
```diff
@@ -1126,18 +1126,25 @@ def test_pdist_empty(self):
...
-                            x.transpose_(-2, -1)
-                        actual = torch.pdist(x, p=p)
-                        expected = brute_pdist(x, p=p)
-                        self.assertEqual(expected.shape, actual.shape)
-                        self.assertTrue(torch.allclose(expected, actual))
+                        test_pdist_single(shape, device, p, trans)
+
+            # do a simplified comparison with big inputs, see:
+            # https://github.com/pytorch/pytorch/issues/15511
+            test_pdist_single((1000, 2), device, 2, False)
```
Comment (ngimel) [inline test/test_torch.py]: 15511 failed for a size of 10000 due to indices computation, not 1000. Not sure though how feasible it is to actually run correctness tests for 10000 though. Also, can we add pdist backward test? I suspect it will fail for smallish sizes because of https://github.com/pytorch/pytorch/blob/2134d8f0838e99794c4fece8599465c377d313e0/aten/src/ATen/native/cuda/DistanceKernel.cu#L193

Comment (gchanan) [reply test/test_torch.py]: I ran the script in the issue and it failed for me at 1000.  I tried 10000 and the test is valid, it just takes much longer (on my machine, ~11 seconds instead of ~1.5 seconds).  I'm actually not sure if that matters or we run the tests in parallel anyway -- @yf225?

Comment (gchanan) [reply test/test_torch.py]: and yes, I'll add a backwards test.

When do you expect that `grid_y` line to cause a problem?  Is it when `grid_y` is > 2^16, so `numel` of around 2^18 (262144), because `block_y` is 4?

File: aten/src/ATen/native/cuda/DistanceKernel.cu (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: Sorry, isn't device_sqrt there?  I'm not sure what you mean.

Comment (erikbrinkman) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: Without testing this, my hunch is that this should be `double` and not whatever precision is actually being used.

```suggestion
  int64_t i = static_cast<int64_t>((n2 - device_sqrt<double>(n2 * n2 - 2 * k - 1)));
```

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 3]
```diff
@@ -106,53 +106,50 @@ __global__ static void pdist_kernel_cuda_impl(scalar_t * result, const scalar_t
...
+  const int stride = blockDim.x * gridDim.x;
 
   if (k >= combs) {
     return;
   }
 
-  float n2 = n - .5;
+  double n2 = n - .5;
   // The -1 accounts for floating point truncation issues
   int64_t i = static_cast<int64_t>((n2 - device_sqrt<scalar_t>(n2 * n2 - 2 * k - 1)));
```
Comment (erikbrinkman) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: See previous comment

```suggestion
  int64_t i = static_cast<int64_t>((n2 - device_sqrt<double>(n2 * n2 - 2 * k - 1)));
```

Comment (gchanan) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: ah that makes sense, and probably what @ngimel meant above.

File: aten/src/ATen/native/cuda/DistanceKernel.cu (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: I updated this to do device_sqrt in double as well, thanks.

Comment (ngimel) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: Sorry, html ate `<double>` from my comment

File: test/test_torch.py (hunk 2 above)
Comment (ngimel) [reply test/test_torch.py]: yes, numel of 2^18, but since `numel = n*(n+1)/2` if I'm not mistaken, that corresponds to n of about 720.

**Comment by jacobrgardner** [comment]
@ezyang @gpleiss Yeah, a working batched pdist is a pretty huge speed and stability win for us. On sufficiently large exact GPs, kernel matrix computations account for ~30% of our running time, and pdist essentially erases all of that time over the best way we know how to do it in native pytorch that doesn't also use too much memory. 

With that said, the non-batch version covers at least some of the use cases for us, whereas the version on master currently is essentially unusable due to the size restrictions.

**Comment by gchanan** [comment]
Ok, I added a large backwards test (thanks to @ngimel) and doubled the maximum dim0 size before pdist backwards blows up (probably at the cost of some performance).

Please speak up if you think this shouldn't go in.

**Comment by ngimel** [comment]
> @ezyang @gpleiss Yeah, a working batched pdist is a pretty huge speed and stability win for us. On sufficiently large exact GPs, kernel matrix computations account for ~30% of our running time, 
@gpleiss, @jacobrgardner what is the range of sizes of batched pdist you are interested in? Is it with p=2 or arbitrary p? Are you also interested in gradients for pdist, or you need just forward?

**Comment by jacobrgardner** [comment]
@ngimel Arbitrary but small `p`. Usual use cases for us would be calling pdist on `X = p x n x n` with `p` probably at most a couple hundred and usually less than 10, and `n` all the way up to 15000+. Gradients are definitely a must for us for all use cases.

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 4]
```diff
@@ -130,10 +127,6 @@ struct PDist {
         self_j += m;
         if (self_j == self_end) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: All this pointer maths smells like a direct port from legacy TH

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 5]
```diff
@@ -197,39 +187,18 @@ struct PDist {
...
-
-      const scalar_t * dist_l = dist_start + l * combs;
-      const scalar_t * grad_l = grad_start + l * combs;
-
-      scalar_t * res_l = res_start + l * n * m;
-      scalar_t * res_v = res_l + v * Vec::size();
-
-      while (start != end) {
-        backward_down_column<F>(self_v, res_v, grad_l, dist_l, pvec, n, m, std::min(int(m - (self_v - self_l)), Vec::size()));
+    at::parallel_for(0, m / Vec::size(), internal::GRAIN_SIZE / (8 * n * n), [=, &pvec](int64_t l, int64_t end) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: I'd try to very carefully audit exactly what variables you're capturing here, since it can make it harder for the compiler to inline the code and avoid costly function calls.. Ideally you catch none. pvec is probably better initialized within that function.

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 6]
```diff
@@ -106,53 +106,50 @@ __global__ static void pdist_kernel_cuda_impl(scalar_t * result, const scalar_t
   // This shared memory is significantly larger than necessary, but the
   // assumption is that it's not a bottleneck, and this is simple
   __shared__ scalar_t shared[forward_threads];
-  int lane = threadIdx.y % warpSize;
-  int warp_id = threadIdx.y / warpSize;
+  int lane = threadIdx.x % warpSize;
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: what's the reason for using threadIdx.x instead of .y here?

File: test/test_autograd.py [hunk 7]
```diff
@@ -2226,6 +2226,22 @@ def test_pow_scalar_base(self):
...
+        def func(x):
+            return torch.pdist(x, p=2)
+
+        devices = ['cpu'] if not torch.cuda.is_available() else ['cpu', 'cuda']
+        for device in devices:
+            # shape[0] should be able to be (roughly) arbitrarily large, but the kernel
+            # is currently limited to smaller sizes (see issue above); this is just testing
+            # a floor.
+            shape = (1000, 1)
+            x = torch.randn(shape, device=device).requires_grad_()
```
Comment (cpuhrsch) [inline test/test_autograd.py]: ideally you also test for all relevant dtypes. I think the default tensor type is double, but we mostly use float.

Comment (cpuhrsch) [reply test/test_autograd.py]: It can also be good to compare the output for both types to see if they roughly equal

File: test/test_torch.py [hunk 8]
```diff
@@ -1126,18 +1126,25 @@ def test_pdist_empty(self):
             self.assertEqual(torch.zeros(3, device=device), torch.pdist(x))
 
     def test_pdist_norm(self):
+        def test_pdist_single(shape, device, p, trans):
```
Comment (cpuhrsch) [inline test/test_torch.py]: as above, it's probably good to also grid over the relevant dtypes

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 4 above)
Comment (erikbrinkman) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: As the original author, I was mostly following what I saw / didn't see other better examples of to handle the arbitrary strides. Not that I'll necessarily have time to rework the batched version, but do you have a better example of an ideal way to handle this?

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: I should flagged this as minor. You could rewrite this to do int64_t comparisons i.e. initialize `j = (l * n + j) * m` and then always do the explicit offset calculation. It'll remove the explicit data dependency, which might help performance wise.

Comment (erikbrinkman) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: I guess I always figured it was "better" to work with pointers to view them like iterators, and that perf would be better with pointers rather than mandating 64bit ints. Note, these thoughts are based on no backing data, other than it generally seems bat to tell the compiler I want 64bit int, when really I want memory addresses and there's no reason to be explicit. Something in the same veing as using size_t instead of in int64.

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: That might be true and probably in particular in the past. It does make it harder to keep track on what's going on though. I'd not use these kind of micro optimizations without careful profiling. Having said that, maybe getting rid of this in this case will make it much worse. In general, I'd expect a modern compiler to know what's the best thing to do here. Something that's harder to figure out however are the exact data dependencies, if you use the same pointer everywhere.

File: aten/src/ATen/native/cuda/DistanceKernel.cu (hunk 6 above)
Comment (gchanan) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: Came from the revert.

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 5 above)
Comment (gchanan) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: Came from the revert.

File: test/test_autograd.py (hunk 7 above)
Comment (gchanan) [reply test/test_autograd.py]: We usually only test autograd in double.

File: test/test_torch.py (hunk 8 above)
Comment (gchanan) [reply test/test_torch.py]: done.

File: test/test_torch.py [hunk 9]
```diff
@@ -1126,18 +1126,25 @@ def test_pdist_empty(self):
...
+                x.transpose_(-2, -1)
+            actual = torch.pdist(x, p=p)
+            expected = brute_pdist(x, p=p)
+            self.assertEqual(expected.shape, actual.shape)
+            self.assertTrue(torch.allclose(expected, actual))
+
         devices = ['cpu'] if not torch.cuda.is_available() else ['cpu', 'cuda']
         for device in devices:
-            for shape in [(4, 5), (3, 2), (2, 1), (2, 3, 4)]:
+            for shape in [(4, 5), (3, 2), (2, 1)]:
```
[the remark this answers is not in the document]
Comment (gchanan) [reply test/test_torch.py]: Why slow down the tests though?  There's already a big test below.

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 10]
```diff
@@ -82,9 +83,9 @@ __global__ static void pdist_kernel_cuda_impl(scalar_t * result, const scalar_t
   const int k = blockIdx.x;
   const int stride = blockDim.x;
 
-  float n2 = n - .5;
-  // The -1 accounts for floating point truncation issues
-  int64_t i = static_cast<int64_t>((n2 - device_sqrt<scalar_t>(n2 * n2 - 2 * k - 1)));
+  double n2 = n - .5;
+  // The -1 accounts for floating point truncation issues; done in fp64 so -1 has an effect up to larger numbers.
+  int64_t i = static_cast<int64_t>((n2 - device_sqrt<double>(n2 * n2 - 2 * k - 1)));
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: I'll check; my first thought is it seems unlikely because it's just done once, not in the loop.

File: test/test_torch.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (cpuhrsch) [reply test/test_torch.py]: This tests for different ps though. If it's not large enough it won't hit all the interesting cases.

**Review by cpuhrsch — APPROVED** [review]
